### PR TITLE
gdrive: add support for teamDriveId

### LINKED ISSUE #3015: add support of team drives for GDrive remote
Google TeamDrives are common for sharing data in academia and education as they provide unlimited storage for free. Right now dvc supports only personal googledrive and it is not clear how to configure a remote to a folder inside of the teamdrive to which you have write access to.

I believe that teamdrive support will be very valuable for dvc (at least we use team drives a lot in our lab). Teamdrives are already supported by rclone and other gdrive libs (see https://github.com/gsuitedevs/PyDrive/issues/149 )

### PR DESCRIPTION
**Pull request by efiop**
Fixes #3015

* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [x] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by efiop** [comment]
CC @MaxRis

**Comment by efiop** [comment]
For the record, @MaxRis is working on this feature now. 🙏

**Comment by MaxRis** [comment]
Team Drive shared folder URL might look like following:
```
gdrive://0AIac4JZqHhKmUk9PDA/test
```
where `0AIac4JZqHhKmUk9PDA` is shared drive root folder id obtained from web browser URL (it will look like `https://drive.google.com/drive/folders/0AIac4JZqHhKmUk9PDA`).

Description of changes in this PR:

- DVC installs PyDrive from GitHub's master URL, since package in PIP is outdated and doesn't contain shared team drives support. ( https://github.com/gsuitedevs/PyDrive@master )
- Fixed support of GDrive URLs pointing to directory id.

File: setup.py [hunk 1]
```diff
@@ -95,7 +95,10 @@ def run(self):
 # google-api-python-client is internal dependency of pydrive. After merge of
 # https://github.com/gsuitedevs/PyDrive/pull/180 into pydrive's master,
 # usage of google-api-python-client can be removed from DVC.
-gdrive = ["pydrive==1.3.1", "google-api-python-client>=1.2"]
+gdrive = [
+    "pydrive @ git+https://github.com/gsuitedevs/PyDrive@master#egg=pydrive",
```
Comment (shcheklein) [inline setup.py]: should we update conda? other packages?

File: dvc/remote/gdrive.py [hunk 2]
```diff
@@ -73,6 +74,17 @@ def __init__(self, repo, config):
         super(RemoteGDrive, self).__init__(repo, config)
         self.no_traverse = False
         self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        bucket_case_sensitive = re.search(
```
Comment (shcheklein) [inline dvc/remote/gdrive.py]: I'm not sure what's going on here? Could you elaborate please? Should we put a comment?

File: tests/unit/remote/test_gdrive.py [hunk 3]
```diff
@@ -18,7 +16,7 @@ class Repo(object):
     tmp_dir = ""
 
 
-class TestRemoteGDrive(TestCase):
+class TestRemoteGDrive(object):
```
Comment (shcheklein) [inline tests/unit/remote/test_gdrive.py]: why this change? is it important for this PR?

File: setup.py (hunk 1 above)
Comment (MaxRis) [reply setup.py]: I guess, yes, will check what version of PyDrive is in conda-forge for the moment. No others packages update is needed, I think.

File: dvc/remote/gdrive.py (hunk 2 above)
Comment (MaxRis) [reply dvc/remote/gdrive.py]: further gdrive logic requires top folder id ("bucket") from URL. For example, for gdrive://0AIac4JZqHhKmUk9PDA/test , folder "test" will be created/used from inside folder with id `0AIac4JZqHhKmUk9PDA`. "bucket" can be "root" or unique folder id.

`self.path_info` already contains parsed `config[Config.SECTION_REMOTE_URL]`, but `urlparse` converts whole URL to lower case.

This source code statement searches for already parsed and lower cased "bucket" part of URL with `IGNORECASE` option in original `Config.SECTION_REMOTE_URL` and retrieves original not lower cased "part" of URL.

For example, it searches for `0aiac4jzqhhkmuk9pda` in `gdrive://0AIac4JZqHhKmUk9PDA/test ` and returns `0AIac4JZqHhKmUk9PDA`.

File: tests/unit/remote/test_gdrive.py (hunk 3 above)
Comment (MaxRis) [reply tests/unit/remote/test_gdrive.py]: it was in original Ruslan's PR. Just left it there. Probably, it's not important, but seems it removes not mandatory dependency on `TestCase`.

File: dvc/remote/gdrive.py (hunk 2 above)
Comment (efiop) [reply dvc/remote/gdrive.py]: @MaxRis Maybe it is easier to just parse the url? :)

Comment (efiop) [reply dvc/remote/gdrive.py]: @MaxRis Also, is `0AIac4JZqHhKmUk9PDA` a team drive id? Is that notation compatible with regular non-team drive urls?

Comment (MaxRis) [reply dvc/remote/gdrive.py]: @efiop 

> @MaxRis Maybe it is easier to just parse the url? :)
Will try with `re`

>  Also, is 0AIac4JZqHhKmUk9PDA a team drive id? Is that notation compatible with regular non-team drive urls?

It is the same unique id which has any Gdrive's file or folder. PyDrive internally adds additional API query params to extend query access for shared drives, but for user it is transparent and no difference if file/folder was accessed from usual or shared drive.

Comment (MaxRis) [reply dvc/remote/gdrive.py]: @efiop have updated that

File: dvc/remote/gdrive.py [hunk 4]
```diff
@@ -73,6 +73,10 @@ def __init__(self, repo, config):
         super(RemoteGDrive, self).__init__(repo, config)
         self.no_traverse = False
         self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.bucket = (
+            config[Config.SECTION_REMOTE_URL].split("://")[1].split("/")[0]
```
Comment (shcheklein) [inline dvc/remote/gdrive.py]: @efiop @MaxRis do we have a check somewhere that REMOTE_URL is an actual URL? We need to avoid Index error on `split('://')[1]` even if user puts a wrong URL into it. I suppose it should be somewhere where we determine the remote type by schema, right? Is that one enough to avoid the error here?

Comment (shcheklein) [reply dvc/remote/gdrive.py]: also, just curious - do they actually call the root folder a `bucket`?

Comment (MaxRis) [reply dvc/remote/gdrive.py]: @shcheklein you're right, ` gdrive:test` causes index error. It seems, if URL starts from `gdrive` the instance of RemoteGDrive is instantiated and no additional schema validation is done.


> also, just curious - do they actually call the root folder a bucket?

No, Google Drive doesn't have such terminology as "bucket". "bucket" is used as DVC's specific terminology ( "bucket" is used by others remotes ).

Comment (shcheklein) [reply dvc/remote/gdrive.py]: how do the call "root" in the URL? directory? let's stick with GDrive termniniology. If it's a directory or folder - let's call it `root_folder` or something.

> `starts from gdrive`

so, does it mean that a "local" remote with a path `gdrive/test` will be considered a GDrive as well?

> no additional schema validation is done.

I wonder where should we fix this? per remote, or somewhere globally in the remote type detection code for example? what do you think @efiop ?

Comment (MaxRis) [reply dvc/remote/gdrive.py]: > how do the call "root" in the URL? directory? let's stick with GDrive termniniology. If it's a directory or folder - let's call it root_folder or something.

There is, seems, no special naming. `root_folder` sounds good, but it might sounds similar to already available `root_id` variable, which is id of actual root directory, but not top most.

> so, does it mean that a "local" remote with a path gdrive/test will be considered a GDrive as well?

Sorry, it seems I was wrong, `gdrive/test` is recognized as local path. But still makes sense to add more validation ( `gdrive:test` is passed to remote without additional validation ).

Comment (shcheklein) [reply dvc/remote/gdrive.py]: > ... already available root_id variable, which is id of actual root directory, but not top most.

I saw that you were using "root" before you changed to the `self.bucket`. Is that "root" special name that has `root_id` your are referring to? May be I don't then understand the logic at all. If a shared URL is just a regular one, just with a strange name (id instead of a human readable one), why do we need to change anything at all. Sorry, that I'm asking this. Probably you have answered this already somewhere.

Comment (MaxRis) [reply dvc/remote/gdrive.py]: > Is that "root" special name that has root_id your are referring to?

"root" is the alias used by Google Drive to point to the user's top most directory. DVC references top most directory just by specifying "root" instead of unique id.

`root_id` is unique id of directory with DVC files.

Usually, it will be different directories.

> . If a shared URL is just a regular one, just with a strange name (id instead of a human readable one), why do we need to change anything at all

Following change https://github.com/iterative/dvc/pull/3016/files#diff-21bfd10d01c6715fc8634644730fa720R272 allows usage of ids as top directory. Currently usage of id as top directory in URL is broken. And this also the blocker for shared team drives, since the top directory always is referenced by id for them.

Comment (MaxRis) [reply dvc/remote/gdrive.py]: It might makes sense to rename `root_id` to `dvc_root_id`, or something better.

Comment (shcheklein) [reply dvc/remote/gdrive.py]: thanks, @MaxRis - my understanding is getting better, but I still have some questions. 

yes, `root_id` should be `remote_root_id`. Though, it's already part of the remote class. May be a comment would be enough to understand the difference.

So, does it mean that, before these URLs with Ids were introduced, all URLs always looked like:

`gdrive://root/path/to/remote`?

File: dvc/remote/gdrive.py (hunk 2 above)
Comment (MaxRis) [reply dvc/remote/gdrive.py]: https://github.com/iterative/dvc/pull/3016/commits/b4a663a981a182613f36ba3623ec845155ee7660#diff-21bfd10d01c6715fc8634644730fa720R82

I believe it shoudn't fail under any input with index error or regular expression group not found error.

File: dvc/remote/gdrive.py (hunk 4 above)
Comment (MaxRis) [reply dvc/remote/gdrive.py]: b4a663a#diff-21bfd10d01c6715fc8634644730fa720R82

I believe it shouldn't fail under any input with index error or regular expression group not found error.


> So, does it mean that, before these URLs with Ids were introduced, all URLs always looked like:

gdrive://root/path/to/remote

I've tested such URLs as gdrive://ID` or gdrive://ED/path/to/remote with my initial changes, but it seems support was broken in finally merged PR after huge amount of modifications. My intention is to cover different URLs types with tests (hopefully, it will be not hard to verify that directories are created property at remote).

Comment (shcheklein) [reply dvc/remote/gdrive.py]: ok, got it. Thanks.

File: dvc/remote/gdrive.py [hunk 5]
```diff
@@ -73,6 +74,14 @@ def __init__(self, repo, config):
         super(RemoteGDrive, self).__init__(repo, config)
         self.no_traverse = False
         self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+
+        bucket = re.search(
+            "{}://(.*)".format(self.scheme), config[Config.SECTION_REMOTE_URL]
```
Comment (shcheklein) [inline dvc/remote/gdrive.py]: is it case sensitive/insensitive? does it matter at all?

Comment (shcheklein) [reply dvc/remote/gdrive.py]: I primarily mean `self.scheme` matching

Comment (MaxRis) [reply dvc/remote/gdrive.py]: Sounds reasonable, thanks!

**Comment by MaxRis** [comment]
docs update - https://github.com/iterative/dvc.org/pull/914

**Comment by efiop** [comment]
Thank you so much @MaxRis ! 🙏
